expression/builtin_regexp.go

```diff
@@ -0,0 +1,1400 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"encoding/hex"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"unicode/utf8"
+
+	"github.com/pingcap/tidb/parser/charset"
+	"github.com/pingcap/tidb/parser/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/collate"
+	"github.com/pingcap/tidb/util/set"
+	"github.com/pingcap/tidb/util/stringutil"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+type empty struct{}
+
+const patternIdx = 1
+const regexpLikeMatchTypeIdx = 2
+const regexpSubstrMatchTypeIdx = 4
+const regexpInstrMatchTypeIdx = 5
+const regexpReplaceMatchTypeIdx = 5
+
+// Valid flags in match type
+const (
+	flagI = "i"
+	flagC = "c"
+	flagM = "m"
+	flagS = "s"
+)
+
+const (
+	invalidMatchType    = "Invalid match type"
+	invalidIndex        = "Index out of bounds in regular expression search"
+	invalidReturnOption = "Incorrect arguments to regexp_instr: return_option must be 1 or 0"
+)
+
+var validMatchType = set.NewStringSet(
+	flagI, // Case-insensitive matching
+	flagC, // Case-sensitive matching
+	flagM, // Multiple-line mode
+	flagS, // The . character matches line terminators
+)
+
+type regexpBaseFuncSig struct {
+	baseBuiltinFunc
+	regexpMemorizedSig
+	once sync.Once
+}
+
+// check binary collation, not xxx_bin collation!
+func (re *regexpBaseFuncSig) isBinaryCollation() bool {
+	return re.collation == charset.CollationBin && re.charset == charset.CharsetBin
+}
+
+func (re *regexpBaseFuncSig) clone() *regexpBaseFuncSig {
+	newSig := &regexpBaseFuncSig{regexpMemorizedSig: regexpMemorizedSig{lock: sync.Mutex{}}, once: sync.Once{}}
+	if re.memorizedRegexp != nil {
+		newSig.memorizedRegexp = re.memorizedRegexp.Copy()
```

Comment:
`Copy` is deprecated due to `Regexp` is safe for concurrent use by multiple goroutines except for configuration methods, such as Longest.

Reply:
> `Copy` is deprecated due to `Regexp` is safe for concurrent use by multiple goroutines except for configuration methods, such as Longest.

I will remove it